ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs

```diff
@@ -206,34 +228,46 @@ chainSyncClient tracer cfg btime (ClockSkew maxSkew) getCurrentChain
             ChainSyncClient .  intersectUnchanged varCandidate
         }
 
-    -- One of the points we sent intersected our chain
-    intersectImproved :: TVar m (CandidateState blk hdr)
+    -- One of the points we sent intersected our chain. This intersection
+    -- point will become the new tip of the candidate chain.
+    intersectImproved :: TVar m (CandidateState hdr)
                       -> Point hdr -> Point hdr
                       -> m (Consensus ClientStIdle hdr m)
     intersectImproved varCandidate intersection _theirHead = atomically $ do
-      -- TODO #472
-      curLedger <- ledgerState <$> getCurrentLedger
 
-      CandidateState { candidateChain } <- readTVar varCandidate
+      CandidateState { candidateChain, candidateChainState } <- readTVar varCandidate
       -- Roll back the candidate to the @intersection@.
-      candidateChain' <- case AF.rollback intersection candidateChain of
-        Just c  -> return c
-        -- The @intersection@ is not on the candidate chain, even though we
-        -- sent only points from the candidate chain to find an intersection
-        -- with. The node must have sent us an invalid intersection point.
-        Nothing -> disconnect $ InvalidIntersection intersection
-
-      -- Get the HeaderState corresponding to the point/block/header we rolled
-      -- back to.
-      let candidateHeaderState' =
-            getHeaderStateFor curLedger candidateChain candidateChain'
+      --
+      -- While the primitives in the ChainSync protocol are "roll back", "roll
+      -- forward (apply block)", etc. The /real/ primitive is "switch to
+      -- fork", which means that a roll back is always followed by applying at
+      -- least as many blocks that we rolled back.
```

Comment:
This assumes that chain selection will always prefer a longer chain, which isn't necessarily the case?

Reply:
No, this assumes it will always prefer a chain that is *at least as long* as the current chain. That will be the case, right?